pymtl3/stdlib/rtl/queues.py

```diff
@@ -86,11 +86,298 @@ def construct( s, num_entries=2 ):
+    s.deq   = DeqIfcRTL( EntryType )
+    s.count = OutPort( mk_bits( clog2( num_entries ) ) )
+
+    # Components
+
+    s.ctrl  = PipeQueueCtrlRTL ( num_entries )
+    s.dpath = NormalQueueDpathRTL( EntryType, num_entries )
+
+    # Connect ctrl to data path
+
+    s.connect( s.ctrl.wen,     s.dpath.wen     )
+    s.connect( s.ctrl.waddr,   s.dpath.waddr   )
+    s.connect( s.ctrl.raddr,   s.dpath.raddr   )
+
+    # Connect to interface
+
+    s.connect( s.enq.en,  s.ctrl.enq_en   )
+    s.connect( s.enq.rdy, s.ctrl.enq_rdy  )
+    s.connect( s.deq.en,  s.ctrl.deq_en   )
+    s.connect( s.deq.rdy, s.ctrl.deq_rdy  )
+    s.connect( s.count,   s.ctrl.count    )
+    s.connect( s.enq.msg, s.dpath.enq_msg )
+    s.connect( s.deq.msg, s.dpath.deq_msg )
+
+  # Line trace
+
+  def line_trace( s ):
+    return "{}({}){}".format( s.enq, s.count, s.deq )
+
+#-------------------------------------------------------------------------
+# Ctrl and Dpath for BypassQueue
+#-------------------------------------------------------------------------
+
+class BypassQueueDpathRTL( Component ):
+
+  def construct( s, EntryType, num_entries=2 ):
+
+    # Interface
+
+    s.enq_msg =  InPort( EntryType )
+    s.deq_msg = OutPort( EntryType )
+
+    s.wen     = InPort( Bits1 )
+    s.waddr   = InPort( mk_bits( clog2( num_entries ) ) )
+    s.raddr   = InPort( mk_bits( clog2( num_entries ) ) )
+    s.mux_sel = InPort( Bits1 )
+
+    # Component
+
+    s.queue = RegisterFile( EntryType, num_entries )(
+      raddr = { 0: s.raddr   },
+      wen   = { 0: s.wen     },
+      waddr = { 0: s.waddr   },
+      wdata = { 0: s.enq_msg },
+    )
+
+    s.mux = Mux( EntryType, 2 )(
+      sel = s.mux_sel,
+      in_ = { 0: s.queue.rdata[0], 1: s.enq_msg },
+      out = s.deq_msg,
+    )
+
+class BypassQueueCtrlRTL( Component ):
+
+  def construct( s, num_entries=2 ):
+
+    # Constants
+
+    s.num_entries = num_entries
+    s.last_idx    = num_entries - 1
+    addr_nbits    = clog2( num_entries   )
+    count_nbits   = clog2( num_entries+1 )
+    PtrType       = mk_bits( addr_nbits  )
+    CountType     = mk_bits( count_nbits )
+
+    # Interface
+
+    s.enq_en  = InPort ( Bits1   )
+    s.enq_rdy = OutPort( Bits1   )
+    s.deq_en  = InPort ( Bits1   )
+    s.deq_rdy = OutPort( Bits1   )
+    s.count   = OutPort( PtrType )
+
+    s.wen     = OutPort( Bits1   )
+    s.waddr   = OutPort( PtrType )
+    s.raddr   = OutPort( PtrType )
+    s.mux_sel = OutPort( Bits1   )
+
+    # Registers
+
+    s.head = Wire( PtrType )
+    s.tail = Wire( PtrType )
+
+    # Wires
+
+    s.enq_xfer  = Wire( Bits1   )
+    s.deq_xfer  = Wire( Bits1   )
+    s.head_next = Wire( PtrType )
+    s.tail_next = Wire( PtrType )
+
+    # Connections
+
+    s.connect( s.wen,   s.enq_xfer )
+    s.connect( s.waddr, s.tail     )
+    s.connect( s.raddr, s.head     )
+
+    @s.update
+    def up_enq_rdy():
+      if ~s.reset:
+        s.enq_rdy = b1(1) if s.count < s.num_entries else b1(0)
```

Comment:
how about ```s.enq_rdy = s.count < s.num_entries```?

Reply:
does boolean expression returns Bits1?

Reply:
I think yes according to verilog conventions and I implemented Bits in that way https://github.com/cornell-brg/pymtl3/blob/master/pymtl3/datatypes/PythonBits.py#L125